api/src/workers/base-event-handler.ts

```diff
@@ -1,129 +1,95 @@
-      } catch (e) {
-        console.error(
-          `error retrieving events for block range fromBlock=${fromBlock} toBlock=${toBlock}`,
-          e
-        );
-        continue;
-      }
-
-      for (let i = 0; i < eventList.length; i++) {
-        console.log(
-          "block_height =",
-          getEventsResult.events[i].blockHeight,
-          " payload =",
-          eventList[i].data
-        );
-        const blockHeight = getEventsResult.events[i].blockHeight;
-        await this.processAndCatchEvent(blockHeight, eventList, i);
-      }
-
-      // update cursor
-      blockCursor = await this.blockCursorService.updateBlockCursorById(
-        blockCursor.id,
-        toBlock
-      );
+    if (!cursors || !cursors.length) {
+      throw new Error("Could not get block cursor from database.");
     }
-  }
 
-  private async processAndCatchEvent(blockHeight, eventList, i: number) {
-    try {
-      await this.onEvent({ blockHeight }, eventList[i].data);
-    } catch (e) {
-      // If we get an error, we're just continuing the loop for now, but they in a production graded app they should
-      // be handled accordingly
-      console.error(`error processing event block_height=${blockHeight}`, e);
-    }
+    cursors.forEach(async ({ cursor, eventName }) => {
+      let blockCursor = await cursor;
+
+      const poll = async () => {
+        let fromBlock, toBlock;
+
+        try {
+          // Calculate block range we're looking at
+          ({ fromBlock, toBlock } = await this.getBlockRange(blockCursor));
+        } catch (e) {
+          console.warn("Error retrieving block range:", e);
+        }
+
+        try {
+          const result = await send([getEvents(eventName, fromBlock, toBlock)]);
+          const decoded = await fcl.decode(result);
+
+          if (decoded.length) {
+            decoded.forEach(async (event) => await this.onEvent(event));
+            // Record the last block where we saw the event(s) we're interested in.
+            blockCursor = await this.blockCursorService.updateBlockCursorById(
+              blockCursor.id,
+              toBlock
+            );
+          }
+        } catch (e) {
+          console.error(
+            `Error retrieving events for block range fromBlock=${fromBlock} toBlock=${toBlock}`,
+            e
+          );
+        }
+
+        setTimeout(poll, this.stepTimeMs);
+      };
+      poll();
+    });
   }
 
-  abstract onEvent(details: EventDetails, payload: any): Promise<void>;
+  abstract onEvent(event: any): Promise<void>;
 
   private async getBlockRange(currentBlockCursor: BlockCursor) {
     const latestBlockHeight = await this.flowService.getLatestBlockHeight();
     const fromBlock = currentBlockCursor.current_block_height;
-    let toBlock = currentBlockCursor.current_block_height + this.stepSize;
-    if (toBlock > latestBlockHeight.height) {
-      toBlock = latestBlockHeight.height;
-    }
+    let toBlock = latestBlockHeight.height;
```

Comment:
I think it's probably a good idea to keep the `stepSize` parameter. If somebody stops their server for 1000 blocks and restarts it, this code will try to catch query all 1000 blocks at once. This will fail and the server enters a crash loop.